File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        # [123]
        # run input 23
        # clip 1
        # 23 append 1
        # 32 append 1
        # append 1 at the end
        # 123
        
        result = []
        
        if (len(nums)==1):
            return [nums[:]]
        
        for i in range(len(nums)):
            n = nums.pop(0)
            perms = self.permute(nums)
            
            for perm in perms:
                perm.append(n)
            
            result.extend(perms)
            nums.append(n)
        return result
        
        

        

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        
        List<List<Integer>> result = new ArrayList<>();

        backtrack(result, new ArrayList<>(), nums);

        return result;
    }

    void backtrack(List<List<Integer>>  result, ArrayList<Integer> tempList, int[] nums) {

        // base case
        if(tempList.size() == nums.length) {
            result.add(new ArrayList(tempList));
            return;
        }

        for(int num : nums) {
            if(tempList.contains(num))
                continue;
            tempList.add(num);
            backtrack(result, tempList, nums);
            tempList.remove(tempList.size()-1);
        }

    }
}